### PR TITLE
ffi: cleanup structseq

### PR DESCRIPTION
**Pull request by deantvv**
For #1289

### CODE REVIEW

File: src/ffi/structseq.rs [hunk 1]
```diff
@@ -18,9 +19,41 @@ pub struct PyStructSequence_Desc {
...
+    pub fn PyStructSequence_InitType(_type: *mut PyTypeObject, desc: *mut PyStructSequence_Desc);
+
+    #[cfg(not(Py_LIMITED_API))]
+    #[cfg_attr(PyPy, link_name = "PyPyStructSequence_InitType2")]
+    pub fn PyStructSequence_InitType2(
+        _type: *mut PyTypeObject,
+        desc: *mut PyStructSequence_Desc,
+    ) -> c_int;
+
     pub fn PyStructSequence_NewType(desc: *mut PyStructSequence_Desc) -> *mut PyTypeObject;
```
Comment (davidhewitt) [inline src/ffi/structseq.rs:36]:
```suggestion
    #[cfg(not(PyPy))]
    #[cfg_attr(docsrs, doc(cfg(not(PyPy))))]
    pub fn PyStructSequence_NewType(desc: *mut PyStructSequence_Desc) -> *mut PyTypeObject;
```

File: src/ffi/structseq.rs [hunk 2]
```diff
@@ -18,9 +19,41 @@ pub struct PyStructSequence_Desc {
...
+
+#[cfg(not(any(Py_LIMITED_API, PyPy)))]
+#[inline]
+pub unsafe fn PyStructSequence_GET_ITEM(op: *mut PyObject, i: Py_ssize_t) -> *mut PyObject {
+    PyTuple_GET_ITEM(op, i)
+}
+
+extern "C" {
     pub fn PyStructSequence_SetItem(arg1: *mut PyObject, arg2: Py_ssize_t, arg3: *mut PyObject);
     pub fn PyStructSequence_GetItem(arg1: *mut PyObject, arg2: Py_ssize_t) -> *mut PyObject;
```
Comment (davidhewitt) [inline src/ffi/structseq.rs:58]:
```suggestion
    #[cfg(not(PyPy))]
    #[cfg_attr(docsrs, doc(cfg(not(PyPy))))]
    pub fn PyStructSequence_SetItem(arg1: *mut PyObject, arg2: Py_ssize_t, arg3: *mut PyObject);
    #[cfg(not(PyPy))]
    #[cfg_attr(docsrs, doc(cfg(not(PyPy))))]
    pub fn PyStructSequence_GetItem(arg1: *mut PyObject, arg2: Py_ssize_t) -> *mut PyObject;
```

File: src/ffi/structseq.rs [hunk 3]
```diff
@@ -1,5 +1,6 @@
 use crate::ffi::object::{PyObject, PyTypeObject};
 use crate::ffi::pyport::Py_ssize_t;
+use crate::ffi::tupleobject::{PyTupleObject, PyTuple_GET_ITEM, PyTuple_SET_ITEM};
```
Comment (davidhewitt) [inline src/ffi/structseq.rs:3]:
```suggestion
#[cfg(not(Py_LIMITED_API))]
use crate::ffi::tupleobject::{PyTupleObject, PyTuple_GET_ITEM, PyTuple_SET_ITEM};
```

**Review by davidhewitt — COMMENTED** [review]
Thanks, looks great! Just a few tweaks needed before merging...

File: src/ffi/structseq.rs (hunk 1 above)
Comment (deantvv) [reply src/ffi/structseq.rs:39]: Thanks for always catching these!

File: src/ffi/structseq.rs (hunk 2 above)
Comment (deantvv) [reply src/ffi/structseq.rs]: Thanks!

File: src/ffi/structseq.rs (hunk 3 above)
Comment (davidhewitt) [reply src/ffi/structseq.rs:4]: Hmm, this import is still failing on PyPy. It's probably tidier to remove this and use the definitions inline below?

File: src/ffi/structseq.rs [hunk 4]
```diff
@@ -18,9 +20,48 @@ pub struct PyStructSequence_Desc {
...
+
+    #[cfg(not(PyPy))]
+    #[cfg_attr(docsrs, doc(cfg(not(PyPy))))]
     pub fn PyStructSequence_NewType(desc: *mut PyStructSequence_Desc) -> *mut PyTypeObject;
+    #[cfg_attr(PyPy, link_name = "PyPyStructSequence_New")]
     pub fn PyStructSequence_New(_type: *mut PyTypeObject) -> *mut PyObject;
+}
+
+#[cfg(not(Py_LIMITED_API))]
+pub type PyStructSequence = PyTupleObject;
```
Comment (davidhewitt) [inline src/ffi/structseq.rs:45]:
```suggestion
pub type PyStructSequence = crate::ffi::PyTupleObject;
```

File: src/ffi/structseq.rs [hunk 5]
```diff
@@ -18,9 +20,48 @@ pub struct PyStructSequence_Desc {
...
+
+#[cfg(not(any(Py_LIMITED_API, PyPy)))]
+#[inline]
+pub unsafe fn PyStructSequence_SET_ITEM(op: *mut PyObject, i: Py_ssize_t, v: *mut PyObject) {
+    PyTuple_SET_ITEM(op, i, v)
```
Comment (davidhewitt) [inline src/ffi/structseq.rs:50]:
```suggestion
    crate::ffi::PyTuple_SET_ITEM(op, i, v)
```

File: src/ffi/structseq.rs [hunk 6]
```diff
@@ -18,9 +20,48 @@ pub struct PyStructSequence_Desc {
...
+}
+
+#[cfg(not(any(Py_LIMITED_API, PyPy)))]
+#[inline]
+pub unsafe fn PyStructSequence_GET_ITEM(op: *mut PyObject, i: Py_ssize_t) -> *mut PyObject {
+    PyTuple_GET_ITEM(op, i)
```
Comment (davidhewitt) [inline src/ffi/structseq.rs:56]:
```suggestion
    crate::ffi::PyTuple_GET_ITEM(op, i)
```

**Review by davidhewitt — COMMENTED** [review]
Hmmm looks like we've still got a bit of awkwardness on PyPy...

**Comment by deantvv** [comment]
> 
> 
> Hmmm looks like we've still got a bit of awkwardness on PyPy...

Resolved. Hope this time we can get a green light in CI.

**Review by davidhewitt — APPROVED** [review]
Hooray looks like we're finally there! Thanks again! 🚀

**Pull request merged**